File: windows/Tracenium.AgentTray/ConsentWindow.cs
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

namespace Tracenium.AgentTray;

/// <summary>
/// El diálogo de consentimiento de ADR-0012, con la cara de Tracenium.
///
/// ── Por qué NO es un MessageBox ─────────────────────────────────────
///
/// Lo era. Un MessageBox sale por encima de todo sin depender de que la
/// bandeja tenga el foco, y es la ventana que Windows ya sabe colocar bien en
/// cualquier DPI y cualquier tema. Por eso se eligió.
///
/// Lo que se perdía a cambio es lo que este diálogo necesita más que ningún
/// otro: **que se reconozca de un vistazo quién lo muestra**. Un aviso que
/// aparece de la nada diciendo que alguien quiere ver tu pantalla, con el
/// aspecto exacto de cualquier otro cuadro del sistema, es indistinguible de
/// un intento de estafa — y la reacción sana ante eso es pulsar lo que sea
/// para que desaparezca. Nombrar el producto en el texto ayuda; enseñarlo con
/// su cromo y su logo, más.
///
/// Todo lo que el MessageBox hacía bien se conserva a propósito:
///
///  * **Denegar es el botón por defecto** (Return) y Escape también deniega.
///    En un diálogo que concede acceso a la pantalla de alguien, la opción de
///    reposo no puede ser la que concede: un Return distraído no debe regalar
///    el control del equipo.
///  * **Modal de verdad**, no una ventana que se pueda dejar de lado.
///  * **TopMost**, porque la persona puede estar en pantalla completa.
///
/// Y se añade lo que no había: la cuenta atrás. El plazo existía —vencerlo
/// cuenta como negativa— pero no se veía, así que la persona no sabía que
/// estaba decidiendo contra reloj.
/// </summary>
internal sealed class ConsentWindow : Form
{
    private const int DialogWidth = 480;

    private static readonly Color Chrome = Color.FromArgb(0x22, 0x28, 0x31);
    private static readonly Color ChromeControl = Color.FromArgb(0x2A, 0x26, 0x20);
    private static readonly Color Cyan = Color.FromArgb(0x8F, 0xFD, 0xFF);
    private static readonly Color Amber = Color.FromArgb(0xF4, 0xD3, 0x7D);
    private static readonly Color Teal = Color.FromArgb(0x3C, 0x7C, 0x7C);
    private static readonly Color Ink = Color.FromArgb(0x1C, 0x20, 0x27);
    private static readonly Color InkSoft = Color.FromArgb(0x5C, 0x64, 0x6E);
    private static readonly Color InkOnAmber = Color.FromArgb(0x3A, 0x2D, 0x07);

    private readonly ConsentRequest _request;
    private readonly bool _control;
    private readonly Label _expiry = new();
    private readonly System.Windows.Forms.Timer _ticker = new() { Interval = 1000 };

    /// <summary>true = la persona concedió.</summary>
    public bool Approved { get; private set; }

    public ConsentWindow(ConsentRequest request)
    {
        _request = request;
        _control = string.Equals(request.Kind, "control", StringComparison.OrdinalIgnoreCase);

        FormBorderStyle = FormBorderStyle.FixedDialog;
        StartPosition = FormStartPosition.CenterScreen;
        ControlBox = false;
        ShowInTaskbar = false;
        TopMost = true;
        BackColor = Color.White;
        Width = DialogWidth;
        AutoScaleMode = AutoScaleMode.Dpi;

        var y = BuildHeader();
        y = BuildBody(y);
        BuildFooter(y);

        _ticker.Tick += (_, _) => Tick();
        Shown += (_, _) => { Tick(); _ticker.Start(); };
        FormClosed += (_, _) => _ticker.Stop();
    }

    /// <summary>Banda oscura con el logo y el eslogan: lo PRIMERO que se ve es
    /// de quién viene el aviso.</summary>
    private int BuildHeader()
    {
        var bar = new Panel
        {
            Location = new Point(0, 0),
            Size = new Size(DialogWidth, 46),
            BackColor = _control ? ChromeControl : Chrome
        };

        var logo = new PictureBox
        {
            Image = BrandAssets.LoadLogoOrNull(),
            SizeMode = PictureBoxSizeMode.Zoom,
            Size = new Size(22, 22),
            Location = new Point(18, 12),
            BackColor = Color.Transparent
        };

        var name = new Label
        {
            Text = "Tracenium",
            AutoSize = true,
            UseMnemonic = false,
            ForeColor = Color.FromArgb(0xF2, 0xF4, 0xF7),
            Location = new Point(50, 15),
            Font = new Font(Font.FontFamily, 9.5f, FontStyle.Bold)
        };

        // El eslogan, con el "&" en cian. Tres etiquetas porque WinForms no
        // pinta parte de un Label en otro color — y las tres con
        // UseMnemonic=false, porque si no el "&" desaparece: es el prefijo de
        // tecla de acceso. Ese mismo fallo tenía la ventana de estado.
        var sloganLeft = new Label
        {
            Text = BrandAssets.SloganLeft,
            AutoSize = true,
            UseMnemonic = false,
            ForeColor = Color.FromArgb(0x9A, 0xA3, 0xAE),
            Font = new Font(Font.FontFamily, 8f)
        };
        var sloganAccent = new Label
        {
            Text = BrandAssets.SloganAccent,
            AutoSize = true,
            UseMnemonic = false,
            ForeColor = Cyan,
            Font = new Font(Font.FontFamily, 8f, FontStyle.Bold)
        };
        var sloganRight = new Label
        {
            Text = BrandAssets.SloganRight,
            AutoSize = true,
            UseMnemonic = false,
            ForeColor = Color.FromArgb(0x9A, 0xA3, 0xAE),
            Font = new Font(Font.FontFamily, 8f)
        };

        bar.Controls.Add(logo);
        bar.Controls.Add(name);
        bar.Controls.Add(sloganLeft);
        bar.Controls.Add(sloganAccent);
        bar.Controls.Add(sloganRight);

        // Las etiquetas son AutoSize: sus anchos sólo se conocen tras medir el
        // texto, así que se encadenan de derecha a izquierda al mostrarse.
        bar.Resize += (_, _) => LayoutSlogan(bar, sloganLeft, sloganAccent, sloganRight);
        Shown += (_, _) => LayoutSlogan(bar, sloganLeft, sloganAccent, sloganRight);

        if (_control)
        {
            bar.Controls.Add(new Panel { Dock = DockStyle.Bottom, Height = 2, BackColor = Amber });
        }

        Controls.Add(bar);
        return bar.Bottom;
    }

    private static void LayoutSlogan(Control bar, Control left, Control accent, Control right)
    {
        var y = (bar.Height - left.Height) / 2;
        right.Location = new Point(bar.Width - 18 - right.Width, y);
        accent.Location = new Point(right.Left - accent.Width, y);
        left.Location = new Point(accent.Left - left.Width, y);
    }

    private int BuildBody(int top)
    {
        var y = top + 20;

        var title = new Label
        {
            Text = _request.Title,
            AutoSize = false,
            UseMnemonic = false,
            ForeColor = Ink,
            Location = new Point(24, y),
            Width = DialogWidth - 48,
            Font = new Font(Font.FontFamily, 12f, FontStyle.Bold)
        };
        title.Height = TextRenderer.MeasureText(title.Text, title.Font,
            new Size(title.Width, 0), TextFormatFlags.WordBreak).Height + 4;
        Controls.Add(title);
        y = title.Bottom + 14;

        // Las líneas vienen YA PARTIDAS del agente (`consent-text.ts`), que es
        // donde vive la redacción para las tres plataformas. Aquí se
        // presentan, no se reescriben: una copia distinta por sistema es una
        // copia que se desincroniza.
        foreach (var line in _request.Lines)
        {
            if (string.IsNullOrWhiteSpace(line)) continue;

            var bullet = new Label
            {
                Text = "•",
                AutoSize = true,
                UseMnemonic = false,
                ForeColor = _control ? Amber : Teal,
                Location = new Point(24, y),
                Font = new Font(Font.FontFamily, 10f, FontStyle.Bold)
            };
            var body = new Label
            {
                Text = line,
                AutoSize = false,
                UseMnemonic = false,
                ForeColor = Ink,
                Location = new Point(44, y),
                Width = DialogWidth - 68,
                Font = new Font(Font.FontFamily, 9.5f)
            };
            body.Height = TextRenderer.MeasureText(body.Text, body.Font,
                new Size(body.Width, 0), TextFormatFlags.WordBreak).Height + 2;

            Controls.Add(bullet);
            Controls.Add(body);
            y = body.Bottom + 10;
        }

        return y + 6;
    }

    private void BuildFooter(int top)
    {
        _expiry.AutoSize = true;
        _expiry.UseMnemonic = false;
        _expiry.ForeColor = InkSoft;
        _expiry.Location = new Point(24, top + 12);
        _expiry.Font = new Font(Font.FontFamily, 8f);
        Controls.Add(_expiry);

        var deny = Pill(_request.DenyLabel, Color.White, Ink, Ink, 2);
        var allow = Pill(_request.AllowLabel,
                         _control ? Amber : Color.FromArgb(0xEB, 0xF4, 0xF4),
                         _control ? InkOnAmber : Color.FromArgb(0x2F, 0x60, 0x60),
                         _control ? Color.FromArgb(0xB9, 0x8F, 0x17) : Teal,
                         1);

        deny.Location = new Point(DialogWidth - 24 - deny.Width, top + 6);
        allow.Location = new Point(deny.Left - 10 - allow.Width, top + 6);

        deny.Click += (_, _) => Decide(false);
        allow.Click += (_, _) => Decide(true);

        Controls.Add(deny);
        Controls.Add(allow);

        // ⚠️ Los dos atajos apuntan a DENEGAR. AcceptButton es el de Return —
        // aquí es el de negarse, a propósito— y CancelButton el de Escape: el
        // gesto universal de "sácame de aquí" no puede conceder acceso a la
        // pantalla de nadie.
        AcceptButton = deny;
        CancelButton = deny;

        ClientSize = new Size(DialogWidth, deny.Bottom + 20);
    }

    private Button Pill(string text, Color fill, Color ink, Color border, int borderWidth)
    {
        var b = new Button
        {
            Text = text,
            AutoSize = false,
            UseMnemonic = false,
            Size = new Size(Math.Max(96, TextRenderer.MeasureText(text, Font).Width + 36), 34),
            FlatStyle = FlatStyle.Flat,
            BackColor = fill,
            ForeColor = ink,
            Cursor = Cursors.Hand,
            Font = new Font(Font.FontFamily, 9.5f, FontStyle.Bold)
        };
        b.FlatAppearance.BorderColor = border;
        b.FlatAppearance.BorderSize = borderWidth;
        return b;
    }

    private void Decide(bool approved)
    {
        Approved = approved;
        DialogResult = approved ? DialogResult.Yes : DialogResult.No;
        Close();
    }

    private void Tick()
    {
        if (_request.ExpiresAtUtc is not DateTime expiry)
        {
            _expiry.Text = "";
            return;
        }
        var left = (int)Math.Round((expiry - DateTime.UtcNow).TotalSeconds);
        if (left <= 0)
        {
            // El plazo venció. No se concede nada: el agente ya lo cuenta como
            // negativa, y dejar la ventana abierta invitaría a contestar a algo
            // que ya no escucha nadie.
            _expiry.Text = "Expired";
            Decide(false);
            return;
        }
        _expiry.Text = $"Expires in {left} s";
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        base.OnPaint(e);
        e.Graphics.SmoothingMode = SmoothingMode.AntiAlias;
    }
}

File: macos/TraceniumAgentStatus/Sources/ConsentWindow.swift
import AppKit

/// El diálogo de consentimiento de ADR-0012, con la cara de Tracenium.
///
/// ── Por qué NO es un `NSAlert` ──────────────────────────────────────
///
/// Lo era. Un `NSAlert` es la ventana que macOS ya sabe colocar bien en
/// cualquier DPI y cualquier tema, y por eso se eligió: sale por encima de
/// todo sin depender de que la bandeja tenga el foco.
///
/// Lo que se perdía a cambio es lo que este diálogo necesita más que ningún
/// otro: **que se reconozca de un vistazo quién lo muestra**. Un aviso que
/// aparece de la nada diciendo que alguien quiere ver tu pantalla, con el
/// aspecto exacto de cualquier otro cuadro del sistema, es indistinguible de
/// un intento de estafa — y la reacción sana ante eso es pulsar lo que sea
/// para que desaparezca. Nombrar el producto en el texto ayuda; enseñarlo
/// con su cromo y su logo, más.
///
/// Todo lo que el `NSAlert` hacía bien se conserva aquí a propósito:
///
///  * **Denegar es el botón por defecto** (Return) y Escape también deniega.
///    En un diálogo que concede acceso a la pantalla de alguien, la opción de
///    reposo no puede ser la que concede.
///  * **Modal de verdad** (`NSApp.runModal`), no una ventana que se pueda
///    dejar de lado y olvidar.
///  * **El icono del bundle firmado**: el texto lo copia cualquiera, el icono
///    que macOS asocia al bundle no. Es la parte que la persona puede comparar
///    con lo que ve en su barra de menús.
///
/// Y se añade lo que no había: la cuenta atrás. El plazo existía —vencerlo
/// cuenta como negativa— pero no se veía, así que la persona no sabía que
/// estaba decidiendo contra reloj.
final class ConsentWindow {

    private let request: ConsentRequest
    private let control: Bool
    private let window: NSWindow
    private let expiryLabel = NSTextField(labelWithString: "")
    private var timer: Timer?
    private var approved = false

    // Marca. Los mismos valores que la franja y que el header del popover.
    private static let chrome = NSColor(srgbRed: 0x22/255.0, green: 0x28/255.0, blue: 0x31/255.0, alpha: 1)
    private static let chromeControl = NSColor(srgbRed: 0x2A/255.0, green: 0x26/255.0, blue: 0x20/255.0, alpha: 1)
    private static let cyan = NSColor(srgbRed: 0x8F/255.0, green: 0xFD/255.0, blue: 0xFF/255.0, alpha: 1)
    private static let amber = NSColor(srgbRed: 0xF4/255.0, green: 0xD3/255.0, blue: 0x7D/255.0, alpha: 1)
    private static let teal = NSColor(srgbRed: 0x3C/255.0, green: 0x7C/255.0, blue: 0x7C/255.0, alpha: 1)
    private static let ink = NSColor(srgbRed: 0x1C/255.0, green: 0x20/255.0, blue: 0x27/255.0, alpha: 1)
    private static let inkSoft = NSColor(srgbRed: 0x5C/255.0, green: 0x64/255.0, blue: 0x6E/255.0, alpha: 1)

    private static let width: CGFloat = 480

    init(request: ConsentRequest, control: Bool) {
        self.request = request
        self.control = control
        window = NSWindow(
            contentRect: NSRect(x: 0, y: 0, width: Self.width, height: 100),
            styleMask: [.titled, .fullSizeContentView],
            backing: .buffered,
            defer: false
        )
        build()
    }

    /// Bloquea hasta que la persona decide. `true` = concedió.
    func runModal() -> Bool {
        window.center()
        window.makeKeyAndOrderFront(nil)
        startCountdown()
        NSApp.runModal(for: window)
        timer?.invalidate()
        window.orderOut(nil)
        return approved
    }

    /// El contenido montado, para poder renderizarlo en un test sin abrir la
    /// ventana modal (que en un test colgaría el `runModal`).
    var contentViewForTests: NSView? { window.contentView }

    // MARK: - Construcción

    private func build() {
        window.titleVisibility = .hidden
        window.titlebarAppearsTransparent = true
        window.isMovableByWindowBackground = true
        window.standardWindowButton(.closeButton)?.isHidden = true
        window.standardWindowButton(.miniaturizeButton)?.isHidden = true
        window.standardWindowButton(.zoomButton)?.isHidden = true
        window.backgroundColor = .white
        window.level = .modalPanel

        let root = NSStackView()
        root.orientation = .vertical
        root.alignment = .leading
        root.spacing = 0
        root.translatesAutoresizingMaskIntoConstraints = false

        root.addArrangedSubview(header())
        root.addArrangedSubview(body())
        root.addArrangedSubview(footer())

        let content = NSView()
        content.wantsLayer = true
        content.layer?.backgroundColor = NSColor.white.cgColor
        content.addSubview(root)
        NSLayoutConstraint.activate([
            root.leadingAnchor.constraint(equalTo: content.leadingAnchor),
            root.trailingAnchor.constraint(equalTo: content.trailingAnchor),
            root.topAnchor.constraint(equalTo: content.topAnchor),
            root.bottomAnchor.constraint(equalTo: content.bottomAnchor),
            content.widthAnchor.constraint(equalToConstant: Self.width),
        ])
        window.contentView = content
        window.layoutIfNeeded()
        window.setContentSize(content.fittingSize)
    }

    /// Banda oscura con el logo y el eslogan. Lo PRIMERO que se ve es de
    /// quién viene el aviso.
    private func header() -> NSView {
        let bar = NSView()
        bar.wantsLayer = true
        bar.layer?.backgroundColor = (control ? Self.chromeControl : Self.chrome).cgColor
        bar.translatesAutoresizingMaskIntoConstraints = false

        let logo = NSImageView()
        logo.imageScaling = .scaleProportionallyUpOrDown
        logo.image = Bundle.main.url(forResource: "tracenium_logo_color", withExtension: "png")
            .flatMap { NSImage(contentsOf: $0) }
        logo.translatesAutoresizingMaskIntoConstraints = false

        let name = NSTextField(labelWithString: "Tracenium")
        name.font = .systemFont(ofSize: 13, weight: .bold)
        name.textColor = NSColor(srgbRed: 0xF2/255.0, green: 0xF4/255.0, blue: 0xF7/255.0, alpha: 1)
        name.translatesAutoresizingMaskIntoConstraints = false

        let slogan = NSTextField(labelWithString: "")
        slogan.attributedStringValue = Self.sloganAttributed()
        slogan.translatesAutoresizingMaskIntoConstraints = false

        bar.addSubview(logo)
        bar.addSubview(name)
        bar.addSubview(slogan)
        NSLayoutConstraint.activate([
            bar.heightAnchor.constraint(equalToConstant: 46),
            logo.leadingAnchor.constraint(equalTo: bar.leadingAnchor, constant: 18),
            logo.centerYAnchor.constraint(equalTo: bar.centerYAnchor),
            logo.widthAnchor.constraint(equalToConstant: 22),
            logo.heightAnchor.constraint(equalToConstant: 22),
            name.leadingAnchor.constraint(equalTo: logo.trailingAnchor, constant: 10),
            name.centerYAnchor.constraint(equalTo: bar.centerYAnchor),
            slogan.trailingAnchor.constraint(equalTo: bar.trailingAnchor, constant: -18),
            slogan.centerYAnchor.constraint(equalTo: bar.centerYAnchor),
            slogan.leadingAnchor.constraint(greaterThanOrEqualTo: name.trailingAnchor, constant: 12),
        ])
        if control {
            let rule = NSView()
            rule.wantsLayer = true
            rule.layer?.backgroundColor = Self.amber.cgColor
            rule.translatesAutoresizingMaskIntoConstraints = false
            bar.addSubview(rule)
            NSLayoutConstraint.activate([
                rule.leadingAnchor.constraint(equalTo: bar.leadingAnchor),
                rule.trailingAnchor.constraint(equalTo: bar.trailingAnchor),
                rule.bottomAnchor.constraint(equalTo: bar.bottomAnchor),
                rule.heightAnchor.constraint(equalToConstant: 2),
            ])
        }
        return bar
    }

    /// El eslogan de producto, con el "&" en el cian de marca.
    ///
    /// Mismo texto y mismo acento que el header del popover de estado: el
    /// diálogo tiene que parecerse a la ventana que la persona ya ha visto en
    /// su barra de menús, porque parecerse a ella ES la prueba.
    static func sloganAttributed() -> NSAttributedString {
        let font = NSFont.systemFont(ofSize: 11, weight: .regular)
        let base = NSColor(calibratedWhite: 0.68, alpha: 1.0)
        let s = NSMutableAttributedString()
        s.append(NSAttributedString(string: "Endpoint Intelligence ",
                                    attributes: [.font: font, .foregroundColor: base]))
        s.append(NSAttributedString(string: "&",
                                    attributes: [.font: NSFont.systemFont(ofSize: 11, weight: .semibold),
                                                 .foregroundColor: cyan]))
        s.append(NSAttributedString(string: " Compliance Platform",
                                    attributes: [.font: font, .foregroundColor: base]))
        return s
    }

    private func body() -> NSView {
        let stack = NSStackView()
        stack.orientation = .vertical
        stack.alignment = .leading
        stack.spacing = 14
        stack.edgeInsets = NSEdgeInsets(top: 22, left: 24, bottom: 6, right: 24)
        stack.translatesAutoresizingMaskIntoConstraints = false

        let title = NSTextField(wrappingLabelWithString: request.title)
        title.font = .systemFont(ofSize: 17, weight: .bold)
        title.textColor = Self.ink
        title.preferredMaxLayoutWidth = Self.width - 48
        stack.addArrangedSubview(title)

        // Las líneas vienen YA PARTIDAS del agente (`consent-text.ts`), que es
        // donde vive la redacción para las tres plataformas. Aquí se
        // presentan, no se reescriben: una copia distinta por sistema es una
        // copia que se desincroniza.
        for line in request.lines where !line.isEmpty {
            stack.addArrangedSubview(bullet(line))
        }

        NSLayoutConstraint.activate([
            stack.widthAnchor.constraint(equalToConstant: Self.width),
        ])
        return stack
    }

    private func bullet(_ text: String) -> NSView {
        let row = NSStackView()
        row.orientation = .horizontal
        row.alignment = .firstBaseline
        row.spacing = 10

        let dot = NSTextField(labelWithString: "•")
        dot.font = .systemFont(ofSize: 13, weight: .bold)
        dot.textColor = control ? Self.amber : Self.teal

        let label = NSTextField(wrappingLabelWithString: text)
        label.font = .systemFont(ofSize: 13)
        label.textColor = Self.ink
        label.preferredMaxLayoutWidth = Self.width - 72

        row.addArrangedSubview(dot)
        row.addArrangedSubview(label)
        return row
    }

    private func footer() -> NSView {
        let bar = NSStackView()
        bar.orientation = .horizontal
        bar.alignment = .centerY
        bar.spacing = 10
        bar.edgeInsets = NSEdgeInsets(top: 16, left: 24, bottom: 20, right: 24)
        bar.translatesAutoresizingMaskIntoConstraints = false

        expiryLabel.font = .systemFont(ofSize: 11)
        expiryLabel.textColor = Self.inkSoft

        let spacer = NSView()
        spacer.setContentHuggingPriority(.defaultLow, for: .horizontal)

        // Píldoras propias y no el bezel del sistema. Dos razones, y la
        // segunda no es estética: el bezel de macOS pinta el botón gris de
        // siempre, que devuelve el diálogo al aspecto anónimo del que se le
        // quiere sacar; y un control con bezel del sistema no aparece en
        // `cacheDisplay`, así que la captura de revisión salía sin botones —
        // o sea, la pieza que decide el consentimiento era justo la que no se
        // podía revisar.
        let deny = pill(request.denyLabel,
                        fill: .white,
                        text: Self.ink,
                        border: Self.ink,
                        borderWidth: 2,
                        action: #selector(denyTapped))
        deny.keyEquivalent = "\r"   // Return ⇒ DENEGAR. Ver la cabecera.
        // El anillo hace VISIBLE cuál es el de reposo. Sin él, "denegar es el
        // botón por defecto" es una verdad que sólo conoce el código.
        deny.layer?.shadowColor = Self.ink.cgColor
        deny.layer?.shadowOpacity = 0.18
        deny.layer?.shadowRadius = 0
        deny.layer?.shadowOffset = .zero
        deny.layer?.masksToBounds = false

        let allow = pill(request.allowLabel,
                         fill: control ? Self.amber : NSColor(srgbRed: 0xEB/255.0, green: 0xF4/255.0, blue: 0xF4/255.0, alpha: 1),
                         text: control ? NSColor(srgbRed: 0x3A/255.0, green: 0x2D/255.0, blue: 0x07/255.0, alpha: 1) : NSColor(srgbRed: 0x2F/255.0, green: 0x60/255.0, blue: 0x60/255.0, alpha: 1),
                         border: control ? NSColor(srgbRed: 0xB9/255.0, green: 0x8F/255.0, blue: 0x17/255.0, alpha: 1) : Self.teal,
                         borderWidth: 1,
                         action: #selector(allowTapped))

        // Escape también deniega. Un botón no puede llevar dos atajos, así que
        // el de Escape es uno invisible: el gesto universal de "sácame de
        // aquí" no puede acabar concediendo acceso a la pantalla.
        let escapeHatch = NSButton(title: "", target: self, action: #selector(denyTapped))
        escapeHatch.keyEquivalent = "\u{1b}"
        escapeHatch.isHidden = true
        bar.addSubview(escapeHatch)

        bar.addArrangedSubview(expiryLabel)
        bar.addArrangedSubview(spacer)
        bar.addArrangedSubview(allow)
        bar.addArrangedSubview(deny)

        NSLayoutConstraint.activate([bar.widthAnchor.constraint(equalToConstant: Self.width)])
        return bar
    }

    /// Un botón con la forma de la marca: píldora, sin bezel de sistema.
    private func pill(_ title: String,
                      fill: NSColor,
                      text: NSColor,
                      border: NSColor,
                      borderWidth: CGFloat,
                      action: Selector) -> NSButton {
        let b = NSButton(title: title, target: self, action: action)
        b.isBordered = false
        b.wantsLayer = true
        b.layer?.backgroundColor = fill.cgColor
        b.layer?.cornerRadius = 8
        b.layer?.borderWidth = borderWidth
        b.layer?.borderColor = border.cgColor
        b.attributedTitle = NSAttributedString(string: title, attributes: [
            .font: NSFont.systemFont(ofSize: 13, weight: .semibold),
            .foregroundColor: text,
        ])
        b.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            b.heightAnchor.constraint(equalToConstant: 34),
            b.widthAnchor.constraint(greaterThanOrEqualToConstant: 96),
        ])
        return b
    }

    // MARK: - Cuenta atrás

    private func startCountdown() {
        tick()
        guard request.expiresAtUtc != nil else { return }
        let t = Timer(timeInterval: 1.0, repeats: true) { [weak self] _ in self?.tick() }
        RunLoop.main.add(t, forMode: .modalPanel)
        timer = t
    }

    private func tick() {
        guard let expiry = request.expiresAtUtc else {
            expiryLabel.stringValue = ""
            return
        }
        let left = Int(expiry.timeIntervalSinceNow.rounded())
        if left <= 0 {
            // El plazo venció. No se concede nada: el agente ya lo cuenta como
            // negativa, y dejar la ventana abierta invitaría a contestar a algo
            // que ya no escucha nadie.
            expiryLabel.stringValue = "Expired"
            approved = false
            NSApp.stopModal()
            return
        }
        expiryLabel.stringValue = "Expires in \(left) s"
    }

    // MARK: - Decisión

    @objc private func denyTapped() {
        approved = false
        NSApp.stopModal()
    }

    @objc private func allowTapped() {
        approved = true
        NSApp.stopModal()
    }
}
